javascript/ql/src/semmle/javascript/SSA.qll

```diff
@@ -333,25 +333,32 @@ private cached module Internal {
     )
   }
 
+  /**
+   * Gets an SSA definition of `v` that reaches the end of the immediate dominator of `bb`.
+   */
+  pragma[noinline]
+  private SsaDefinition getDefReachingEndOfImmediateDominator(ReachableBasicBlock bb, SsaSourceVariable v) {
+    result = getDefReachingEndOf(bb.getImmediateDominator(), v)
+  }
+
   /**
    * Gets an SSA definition of `v` that reaches the end of basic block `bb`.
    */
   cached SsaDefinition getDefReachingEndOf(ReachableBasicBlock bb, SsaSourceVariable v) {
-    bb.getASuccessor().localIsLiveAtEntry(v) and
-    (
-      exists (int lastRef | lastRef = max(int i | ssaRef(bb, i, v, _)) |
-        result = getLocalDefinition(bb, lastRef, v)
-        or
-        result.definesAt(bb, lastRef, v)
-      )
+    exists (int lastRef | lastRef = max(int i | ssaRef(bb, i, v, _)) |
+      result = getLocalDefinition(bb, lastRef, v)
```

Comment:
Why does the first disjunct not require `liveAtSuccEntry`?

Reply:
As per Esben's comment below, I think the `liveAtSuccEntry` may actually be spurious, but I wouldn't want to change it without redoing the evaluation. As you say, SSA is hairy.

Reply:
I've verified that omitting `liveAtSuccEntry` does not change results on the default project set, but it also doesn't help performance. In fact, I seem to get a slight slowdown, though that may just be a measurement artefact.